pkg/cloudprovider/vsphereparavirtual/route.go

```diff
@@ -0,0 +1,415 @@
+				cpRoute := &cloudprovider.Route{
+					Name:            route.Name,
+					TargetNode:      types.NodeName(nodeName),
+					DestinationCIDR: route.Destination,
+				}
+				routes = append(routes, cpRoute)
+			}
+			// }
+		}
+	}
+	return routes
+}
+
+// CreateRoute implements Routes.CreateRoute
+// Create a RouteSet custom resource for a Node
+func (r *routesProvider) CreateRoute(ctx context.Context, clusterName string, nameHint string, route *cloudprovider.Route) error {
+	nodeName := string(route.TargetNode)
+	klog.V(6).Infof("Creating Route for node %s with hint %s in cluster %s", nodeName, nameHint, clusterName)
+
+	nodeIP, err := r.getNodeIPAddress(nodeName, util.IsIPv4(route.DestinationCIDR))
+	if err != nil {
+		klog.Errorf("getting node %s IP address failed: %v", nodeName, err)
+		return err
+	}
+
+	_, err = r.createRouteSetCR(ctx, clusterName, nameHint, nodeName, route.DestinationCIDR, nodeIP)
+	if err != nil {
+		klog.Errorf("creating RouteSet CR for node %s failed: %s", nodeName, err)
+		return err
+	}
+	return nil
+	// TODO: Get realized state of corresponding static route
+	// return r.checkStaticRouteRealizedState(routeSet.Name)
+}
+
+// createRouteSetCR creates RouteSet CR through RouteSet client
+func (r *routesProvider) createRouteSetCR(ctx context.Context, clusterName string, nameHint string, nodeName string, cidr string, nodeIP string) (*v1alpha1.RouteSet, error) {
+	annotations := map[string]string{
+		AnnotationClusterName: clusterName,
+	}
+	nodeRef := metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        types.UID(nameHint),
+	}
+	r.ownerRefs = append(r.ownerRefs, nodeRef)
+	route := v1alpha1.Route{
+		Name:        r.GetRouteName(nodeName, cidr, clusterName),
+		Destination: cidr,
+		Target:      nodeIP,
+	}
+	routeSetSpec := v1alpha1.RouteSetSpec{
+		Routes: []v1alpha1.Route{
+			route,
+		},
+	}
+	routeSet := &v1alpha1.RouteSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations:     annotations,
+			Name:            nodeName,
+			OwnerReferences: r.ownerRefs,
+			Namespace:       r.namespace,
+		},
+		Spec: routeSetSpec,
+	}
+
+	_, err := r.routeClient.NsxV1alpha1().RouteSets(r.namespace).Create(ctx, routeSet, metav1.CreateOptions{})
+	if err != nil {
+		klog.ErrorS(ErrCreateRouteSet, fmt.Sprintf("%v", err))
+		return nil, err
+	}
+
+	klog.V(6).Infof("Successfully created RouteSet CR for node %s", nodeName)
+	err = r.updateNetworkingCondition(nodeName)
+	if err != nil {
+		klog.V(6).Infof("Error updating NetworkingCondition for node %s, error: %v", types.NodeName(nodeName), err)
+	}
+	return routeSet, nil
+}
+
+// updateNetworkingCondition removes "NodeNetworkUnavailable" condition on the Node once the Route CR is successfully created
+func (r *routesProvider) updateNetworkingCondition(nodeName string) error {
+	klog.V(6).Infof("updating NetworkingCondition for node %s", nodeName)
+
+	nodes, err := r.nodeLister.List(labels.Everything())
```

Comment:
How about using `r.nodeMap` as you have the list already, I'm not exactly sure whether `r.nodeLister.List` will have a extra call to apiserver?

Reply:
I think lister could do something like this
```
ippool, err := c.ippoolLister.IPPools(namespace).Get(name)
```